secure/storage/src/crypto_kv_storage.rs

```diff
@@ -0,0 +1,177 @@
+        // now, and not have to do it later on a rotation.
+        self.create(
+            &Self::get_previous_version_name(key_pair_name),
+            Value::Ed25519PrivateKey(self.get_private_key_for_name(key_pair_name)?),
+            policy,
+        )?;
+
+        Ok(public_key)
+    }
+
+    fn get_public_key_for_name(&self, key_pair_name: &str) -> Result<Ed25519PublicKey, Error> {
+        self.get_private_key_for_name(key_pair_name)
+            .map(|e| e.public_key())
+    }
+
+    fn get_private_key_for_name(&self, key_pair_name: &str) -> Result<Ed25519PrivateKey, Error> {
+        match self.storage.get(key_pair_name)? {
+            Value::Ed25519PrivateKey(private_key) => Ok(private_key),
+            _ => Err(Error::UnexpectedValueType),
+        }
+    }
+
+    fn get_private_key_for_name_and_version(
+        &self,
+        key_pair_name: &str,
+        key_pair_version: Ed25519PublicKey,
+    ) -> Result<Ed25519PrivateKey, Error> {
+        let current_private_key = self.get_private_key_for_name(key_pair_name)?;
+        if current_private_key.public_key().eq(&key_pair_version) {
+            return Ok(current_private_key);
+        }
+
+        let previous_private_key =
+            self.get_private_key_for_name(&Self::get_previous_version_name(key_pair_name))?;
+        if previous_private_key.public_key().eq(&key_pair_version) {
+            return Ok(previous_private_key);
+        }
+
+        Err(Error::KeyVersionNotFound(key_pair_version.to_string()))
+    }
+
+    fn rotate_key_pair(&mut self, key_pair_name: &str) -> Result<Ed25519PublicKey, Error> {
+        match self.storage.get(key_pair_name)? {
+            Value::Ed25519PrivateKey(private_key) => {
+                let (new_private_key, new_public_key) = Self::new_ed25519_key_pair()?;
+                self.storage.set(
+                    &Self::get_previous_version_name(key_pair_name),
+                    Value::Ed25519PrivateKey(private_key),
+                )?;
+                self.storage
+                    .set(key_pair_name, Value::Ed25519PrivateKey(new_private_key))?;
+                Ok(new_public_key)
+            }
+            _ => Err(Error::UnexpectedValueType),
+        }
+    }
+
+    fn sign_message(
+        &mut self,
+        key_pair_name: &str,
+        message: &HashValue,
+    ) -> Result<Ed25519Signature, Error> {
+        let private_key = self.get_private_key_for_name(key_pair_name)?;
+        Ok(private_key.sign_message(message))
+    }
+
+    fn sign_message_using_version(
+        &mut self,
+        key_pair_name: &str,
+        key_pair_version: Ed25519PublicKey,
+        message: &HashValue,
+    ) -> Result<Ed25519Signature, Error> {
+        let private_key =
+            self.get_private_key_for_name_and_version(key_pair_name, key_pair_version)?;
+        Ok(private_key.sign_message(message))
+    }
+}
+
+/// Delegates the KVStorage implementation to the internal storage engine.
+impl KVStorage for CryptoKVStorage {
```

Comment:
Why do we need this?

I imagined something like this:

or some other pattern, this doesn't seem like necessary copy pasta.
trait CryptoKVStorage<T: KVStorage> {
}

Reply:
There is a trade-off between: (1) having CryptoKVStorage be defined as a trait and extending KVStorage; and (2) having CryptoKVStorage be a struct that takes in a KVStorage (as we currently do).

The benefits of option 1 are that we don't need to implement KVStorage (e.g. as above), but the disadvantages are that we: (i) can't have private helper methods (e.g., so we still expose the version_name method function);  and (ii) require each KVStorage implementation (e.g. InMemoryStorage) to implement CryptoKVStorage (so that it can get the default implementation methods), which is really ugly, because technically KVStorage implementations shouldn't know about this.

The only downside of option 2 appears to be having to forward the calls.  But it allows us to hide ugly implementation details and provides a good location for convenience methods to construct CryptoKVStorage engines easily. 

If you still feeling very strongly, push back, and I'll adopt the change. However, personally I'm in favour of option 2.

Reply:
See my comment above, I think we can hide the generics because the generics would be totally encapsulated within the implementation instead of bleeding through to the top level.

Reply:
Thanks! Please see my comment above :)